src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,205 @@
+                                                Encoding fromEncoding, Encoding toEncoding) {
+  Converter to(toEncoding);
+  std::string substitude(to.minSize(), '?');
+  to.setSubstitudeChars(substitude);
+  Converter from(fromEncoding);
+
+  auto limit = source.size() + to.maxSize();
+  auto out = kj::heapArray<kj::byte>(limit);
+  char* target = out.asChars().begin();
+  const char* source_ = source.asChars().begin();
+  UErrorCode status{};
+  ucnv_convertEx(to.conv(), from.conv(), &target, target + limit, &source_, source_ + source.size(),
+                 nullptr, nullptr, nullptr, nullptr, true, true, &status);
+  if (U_SUCCESS(status)) {
+    return out.slice(0, target - out.asChars().begin()).attach(kj::mv(out));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeToUCS2(kj::ArrayPtr<kj::byte> source, Encoding fromEncoding,
+                                               Encoding toEncoding) {
+  UErrorCode status{};
+  const size_t length_in_chars = source.size() * sizeof(UChar);
+  Converter from(fromEncoding);
+  auto out = kj::heapArray<UChar>(source.size());
+  const auto source_ = source.asChars().begin();
+  ucnv_toUChars(from.conv(), out.begin(), length_in_chars, source_, source.size(), &status);
+  if (U_SUCCESS(status)) {
+    return out.asBytes().attach(kj::mv(out));
+  }
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeFromUCS2(kj::ArrayPtr<kj::byte> source,
+                                                 Encoding fromEncoding, Encoding toEncoding) {
+  UErrorCode status{};
+  KJ_STACK_ARRAY(kj::byte, sourcebuf, 1024, 0, 1024);
+  Converter to(toEncoding);
+  std::string substitude(to.minSize(), '?');
+  to.setSubstitudeChars(substitude);
+
+  const size_t length_in_chars = source.size() * sizeof(UChar);
+  sourcebuf.copyFrom(source.slice(0, length_in_chars));
+
+  auto destbuf = kj::heapArray<kj::byte>(length_in_chars);
+  const auto source_ = reinterpret_cast<const UChar*>(sourcebuf.asChars().begin());
+  auto len = ucnv_fromUChars(to.conv(), destbuf.asChars().begin(), length_in_chars, source_,
+                             length_in_chars, &status);
+
+  if (U_SUCCESS(status)) {
+    return destbuf.slice(0, len).attach(kj::mv(destbuf));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeUcs2FromUtf8(kj::ArrayPtr<kj::byte> source,
+                                                     Encoding fromEncoding, Encoding toEncoding) {
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeUtf8FromUcs2(kj::ArrayPtr<kj::byte> source,
+                                                     Encoding fromEncoding, Encoding toEncoding) {
+  return kj::none;
+}
+
+} // namespace
+
+Converter::Converter(Encoding encoding, kj::StringPtr substitude) {
+  UErrorCode status = U_ZERO_ERROR;
+  auto name = getEncodingName(encoding);
+  auto conv = ucnv_open(name, &status);
+  KJ_ASSERT(U_SUCCESS(status));
+  conv_ = kj::Own<UConverter>(conv, ConverterDisposer::INSTANCE);
```

Comment:
kj has a `kj::disposeWith(...)` API that makes this a bit easier. Do a search for it in the code to see examples.